### PR TITLE
Introduce GatewayStream API

### LINKED ISSUE #11706: Introduce external job activation API
**Description**

In order for both the ZPA and ZDP teams to be able to work independently on the job push epic, we should introduce the API for job activation/pushing between the engine and the broker as soon as possible.

Job activation is predicated on workers/streams being available to receive jobs. Since processing in the engine is synchronous, the API should be blocking/synchronous as well.

The engine should be able to query the API for a worker by job type. If there is none, then the job is not activated. If there is one, then the API must return all the properties required to activate the job, and a consumer for activated job.

The engine should activate the job with the given properties (e.g. deadline, worker, etc.). Once activated, it will collect the variables for that job (based on the worker's properties), and pass the job to the consumer. After that, the job should be guaranteed to be activated and locked (based on the deadline).

In the [prototype](https://github.com/camunda/zeebe/pull/11285/files), I didn't differentiate between workers, so I was ignoring all the activation properties. That said, here's what it looked like:

```java
/**
 * Implementations of this interface help the engine decide whether to immediately activate a job
 * when it's made available. This is primarily used to immediately push jobs out to external workers
 * without going through a poll-request-cycle.
 */
@FunctionalInterface
public interface ExternalJobActivator {

  Optional<Handler> activateJob(final DirectBuffer type);

  @FunctionalInterface
  interface Handler {
    void handle(final long key, final JobRecord job);
  }
}
```

We would need to adapt it such that it returns a type which can supply activation properties and the handler/consumer.

Activation properties as far as I can tell are:

- Worker name
- Fetch variables (e.g. the variables to return with the job)
- Deadline/timeout

Initially, this interface can replace the existing `LongPollingJobNotification` and the consumer for job types. The initial implementation can however still do the exact same thing:

- Always return no activation properties
- Send out the job available notification whenever the engine queries the API

Activating the job, collecting the variables, and pushing the job can be done in one or more follow up issues by the ZPA team once this is done.

Regarding observability, I don't think we need any special metrics here just yet.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

Introduces a new platform API to allow record processors to stream data out to Zeebe Gateway instances.

Additionally adds concrete type implementations for the upcoming job streaming feature, and an optional accessor in the `RecordProcessorContext`.

Introducing this API as a first step will allow both the ZPA and ZDP team to work mostly independently with minimal friction, even if it is likely to change in the future.

## Related issues

closes #11706 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
@koevskinikola As we discussed, feel free to move the `JobActivationProperties` and `ActivatedJob` interfaces later into different modules. I have a feeling we will need them elsewhere as well, but let's just start with them there for now.

**Comment by npepinpe** [comment]
I added 4 new commits which actually make use of the `GatewayStreamer` API, but keep the functionality backwards compatible, so it behaves like the old `LongPollingJobNotification`. 

There were some weird things which would need to be addressed in the future:

1. We have to clone the buffer in the `DbJobState` when passing it to the `GatewayStreamer` because of the tests in `ActivatableJobsNotificationTest`. There it uses a static mock, and since the buffer is reused between tests, the verification all fail if we don't clone it. It's not required for production though, so imo we should look into removing this unnecessary clone, either by refactoring the tests or something more clever.
2. It's cumbersome to pass the job streamer from the broker down to the job state. I think this highlights that it's weird that the job state in general should take care of it. I think your team will have a better idea of how to properly refactor this.

Let me know if you'd like me to split those 4 commits into a second PR!

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionFactory.java [hunk 1]
```diff
@@ -170,7 +170,8 @@ List<ZeebePartition> constructPartitions(
               new PartitionProcessingState(owningPartition),
               diskSpaceUsageMonitor,
               gatewayBrokerTransport,
-              topologyManager);
```
Comment (koevskinikola) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionFactory.java:173]: :thought_balloon: Just to clarify why this refactoring is made: we want to make the GatewayStreamer a new top-level service, i.e. the job stream service. We also need to expose the GatewayStreamer for the work you'll do on the broker/gateway side. So for these reasons, we need to move the instantiation up and pass it to the StreamPlatform, correct? 

>It's cumbersome to pass the job streamer from the broker down to the job state. I think this highlights that it's weird that the job state in general should take care of it. I think your team will have a better idea of how to properly refactor this.

I see what you mean when you say that. :smile: I would say that your refactoring makes the code cleaner, but then we get [this](https://github.com/camunda/zeebe/pull/11889/files#diff-71105e453b58e4b990bde2e746beaa2fe0cef0ae1faa45024deea37e8b06374bR78) which feels a bit weird.

I think it's good for now. I will look into what we can improve when I start the task breakdown for the work on the engine side within this epic.

**Review by koevskinikola — APPROVED** [review]
@npepinpe the code looks good. :+1: 

I left some minor review hints, but they shouldn't block you from merging.

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionTransitionContext.java [hunk 2]
```diff
@@ -128,4 +131,6 @@ public interface PartitionTransitionContext extends PartitionContext {
   BackupStore getBackupStore();
 
   void setBackupStore(BackupStore backupStore);
+
+  GatewayStreamer<JobActivationProperties, ActivatedJob> jobStreamer();
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionTransitionContext.java:135]: The only things with setters are mutable members. We can add it if we have a use case to mutate the streamer reference at some point :shrug: See for example `PartitionTransitionContext#getMessagingService()` which also has no setter.

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**